File: lib/rust/graph.rs
use std::cmp::Reverse;
use std::collections::hash_map::Iter;
use std::fmt::Debug;
use std::{
    collections::{BinaryHeap, HashMap, HashSet, VecDeque},
    hash::Hash,
};

use itertools::Itertools;

#[derive(Debug, Clone)]
pub struct Graph<V> {
    adjacency: HashMap<V, HashMap<V, i64>>,
}

impl<V: Eq + Hash> FromIterator<(V, HashSet<V>)> for Graph<V> {
    fn from_iter<T: IntoIterator<Item = (V, HashSet<V>)>>(iter: T) -> Self {
        Graph {
            adjacency: iter
                .into_iter()
                .map(|(vertex, adjacent)| (vertex, adjacent.into_iter().map(|a| (a, 1)).collect()))
                .collect(),
        }
    }
}

impl<V: Clone + Eq + Hash> Graph<V> {
    pub fn empty() -> Self {
        Graph {
            adjacency: HashMap::new(),
        }
    }

    pub fn vertices(&self) -> Vec<V> {
        self.adjacency
            .keys()
            .chain(self.adjacency.values().flat_map(|adj| adj.keys()))
            .unique()
            .cloned()
            .collect()
    }

    pub fn sinks(&self) -> Vec<V> {
        self.vertices()
            .into_iter()
            .filter(|v| self.adjacency.get(v).map_or(true, |adj| adj.is_empty()))
            .collect()
    }

    pub fn iter(&self) -> Iter<'_, V, HashMap<V, i64>> {
        self.adjacency.iter()
    }

    pub fn is_empty(&self) -> bool {
        self.adjacency.is_empty()
    }

    pub fn remove_vertex(&mut self, vertex: &V) {
        self.adjacency.remove(vertex);
        for (_, adj) in self.adjacency.iter_mut() {
            adj.remove(vertex);
        }
    }

    pub fn add_edge(&mut self, from: &V, to: &V, directed: bool) {
        self.add_weighted_edge(from, to, 1, directed)
    }

    pub fn add_weighted_edge(&mut self, from: &V, to: &V, weight: i64, directed: bool) {
        self.adjacency
            .entry(from.clone())
            .or_default()
            .insert(to.clone(), weight);
        if !directed {
            self.adjacency
                .entry(to.clone())
                .or_default()
                .insert(from.clone(), weight);
        }
    }

    pub fn remove_edge(&mut self, from: &V, to: &V, directed: bool) {
        self.adjacency.get_mut(from).unwrap().remove(to);
        if !directed {
            self.adjacency.get_mut(to).unwrap().remove(from);
        }
    }

    pub fn adjacent_vertices(&self, vertex: &V) -> Vec<V> {
        self.adjacency
            .get(vertex)
            .map_or(vec![], |adj| adj.keys().cloned().collect())
    }

    pub fn get_weight(&self, from: &V, to: &V) -> i64 {
        self.adjacency[from][to]
    }

    pub fn floyd_warshall(&self) -> HashMap<(V, V), i32> {
        let vertices = self.vertices();
        let n = vertices.len();
        let mut dist = vec![vec![i32::MAX as i64; n]; n];

        for (u, neighbors) in &self.adjacency {
            let ui = vertices.iter().position(|vert| u == vert).unwrap();
            for (v, weight) in neighbors.iter() {
                let vi = vertices.iter().position(|vert| v == vert).unwrap();
                dist[ui][vi] = *weight;
            }
        }

        for vi in 0..n {
            dist[vi][vi] = 0;
        }

        for k in 0..n {
            for i in 0..n {
                for j in 0..n {
                    if dist[i][j] > dist[i][k] + dist[k][j] {
                        dist[i][j] = dist[i][k] + dist[k][j];
                    }
                }
            }
        }

        (0..n)
            .tuple_combinations()
            .flat_map(|(ui, vi)| {
                [
                    ((vertices[ui].clone(), vertices[vi].clone()), dist[ui][vi] as i32),
                    ((vertices[vi].clone(), vertices[ui].clone()), dist[vi][ui] as i32),
                ]
            })
            .collect()
    }

    pub fn components(&self) -> Vec<HashSet<V>> {
        let mut seen = HashSet::new();
        let mut components = Vec::new();

        let vertices = self.vertices();
        while seen.len() < vertices.len() {
            let vertex = vertices.iter().find(|p| !seen.contains(*p)).unwrap();

            let mut comp = HashSet::new();
            let mut queue = VecDeque::from([vertex.clone()]);
            while let Some(u) = queue.pop_front() {
                comp.insert(u.clone());
                if seen.contains(&u) {
                    continue;
                }

                seen.insert(u.clone());

                for w in self.adjacency[&u].keys() {
                    queue.push_back(w.clone());
                }
            }

            components.push(comp);
        }

        components
    }

    /// Kahn's Algorithm
    pub fn topo_sorting(&self) -> Option<Vec<V>>
    where
        V: Ord,
    {
        // Invert graph to map trg to source nodes
        let mut sources = self.invert().adjacency;

        let mut sorted = Vec::new();
        let mut next: BinaryHeap<_> = sources
            .iter()
            .filter(|(_, srcs)| srcs.is_empty())
            .map(|(trg, _)| Reverse(trg.clone())) // Use reverse to ensure smaller nodes are chosen first
            .collect();

        while let Some(Reverse(node)) = next.pop() {
            if sources.remove(&node).is_some() {
                sorted.push(node.clone());
                for (trg, srcs) in sources.iter_mut() {
                    srcs.remove(&node);
                    if srcs.is_empty() {
                        next.push(Reverse(trg.clone()));
                    }
                }
            }
        }

        if sources.is_empty() {
            Some(sorted)
        } else {
            None
        }
    }

    // Invert graph to map trg to source nodes
    pub fn invert(&self) -> Self {
        self.vertices()
            .into_iter()
            .map(|trg| {
                (
                    trg.clone(),
                    self.adjacency
                        .iter()
                        .filter(|(_, trgs)| trgs.contains_key(&trg))
                        .map(|(src, _)| src.clone())
                        .collect::<HashSet<_>>(),
                )
            })
            .collect()
    }
}

pub trait PathFinding<V: Clone + Eq + Hash + Ord> {
    fn transitions(&mut self, current: &V) -> Vec<(V, i64)>;

    fn astar_no_heuristic(&mut self, start: &V, goal: &V) -> Option<(i64, Vec<V>)> {
        self.astar(start, goal, |_| 0)
    }

    fn astar_many<F, G>(&mut self, start: &V, is_goal: G, heuristic: F) -> Option<(i64, Vec<V>)>
    where
        F: Fn(&V) -> i64,
        G: Fn(&V) -> bool,
    {
        let mut open = BinaryHeap::new();
        let mut closed = HashSet::new();

        open.push((Reverse(0), start.clone(), vec![start.clone()]));
        let mut g_score_map = HashMap::from([(start.clone(), 0)]);

        // Loop until you find the end
        while let Some((_f, current, path)) = open.pop() {
            closed.insert(current.clone());

            let g = g_score_map[&current];

            // Found the goal
            if is_goal(&current) {
                return Some((g, path));
            }

            for (child, weight) in self.transitions(&current) {
                if closed.contains(&child) {
                    continue;
                }

                let child_g = g + weight;
                if *g_score_map.get(&child).unwrap_or(&i64::MAX) <= child_g {
                    continue;
                }
                g_score_map.insert(child.clone(), child_g);

                let child_h = heuristic(&child);
                let child_f = child_g + child_h;

                let mut new_path = path.clone();
                new_path.push(child.clone());
                open.push((Reverse(child_f), child.clone(), new_path))
            }
        }

        None
    }

    fn astar<F>(&mut self, start: &V, goal: &V, heuristic: F) -> Option<(i64, Vec<V>)>
    where
        F: Fn(&V) -> i64,
    {
        self.astar_many(start, |g| g == goal, heuristic)
    }

    /// Shortest paths from start to all other nodes (does not contain total cost of path)
    fn dijkstra(&mut self, start: &V) -> HashMap<V, Vec<V>>
    where
        V: Ord,
    {
        let mut prev = HashMap::new();
        let mut dist = HashMap::from([(start.clone(), 0)]);
        let mut seen_vertices = HashSet::from([start.clone()]);

        let mut queue = BinaryHeap::from([(0, start.clone())]);

        while let Some((_, u)) = queue.pop() {
            for (v, weight) in self.transitions(&u) {
                let alt = dist.get(&u).unwrap_or(&i64::MAX).saturating_add(weight);
                if alt < *dist.get(&v).unwrap_or(&i64::MAX) {
                    *dist.entry(v.clone()).or_default() = alt;
                    prev.entry(v.clone())
                        .and_modify(|e| *e = u.clone())
                        .or_insert(u.clone());
                    queue.push((-alt, v.clone()));
                    seen_vertices.insert(v.clone());
                }
            }
        }

        let mut paths = HashMap::new();
        for v in seen_vertices.into_iter() {
            if v == *start {
                continue;
            }

            let mut curr = v.clone();
            let mut path = vec![v.clone()];
            while curr != *start {
                curr = prev[&curr].clone();
                path.push(curr.clone());
            }
            path.reverse();

            paths.insert(v.clone(), path);
        }

        paths
    }
}

impl<V: Clone + Eq + Hash + Ord> PathFinding<V> for Graph<V> {
    fn transitions(&mut self, current: &V) -> Vec<(V, i64)> {
        if let Some(children) = self.adjacency.get(current) {
            children.iter().map(|(v, w)| (v.clone(), *w)).collect_vec()
        } else {
            vec![]
        }
    }
}

pub struct DynamicGraph<V: Copy> {
    #[allow(clippy::type_complexity)]
    pub adjacent: Box<dyn FnMut(&V) -> Vec<(V, i64)>>,
}

impl<V: Clone + Eq + Hash + Ord + Copy> PathFinding<V> for DynamicGraph<V> {
    fn transitions(&mut self, current: &V) -> Vec<(V, i64)> {
        (self.adjacent)(current)
    }
}

File: lib/zig/graph.zig
const std = @import("std");
const util = @import("util.zig");
const math = std.math;

pub fn Graph(comptime T: type) type {
    return struct {
        const Self = @This();
        adjacency: std.AutoHashMap(T, std.AutoArrayHashMap(T, i32)),
        _vertices: std.AutoArrayHashMap(T, void),
        _allocator: std.mem.Allocator,

        pub fn init(allocator: std.mem.Allocator) Graph(T) {
            return .{
                .adjacency = std.AutoHashMap(T, std.AutoArrayHashMap(T, i32)).init(allocator),
                ._vertices = std.AutoArrayHashMap(T, void).init(allocator),
                ._allocator = allocator,
            };
        }

        pub fn deinit(self: *Self) void {
            self._vertices.deinit();
            var values = self.adjacency.valueIterator();
            while (values.next()) |val| val.deinit();
            self.adjacency.deinit();
        }

        pub fn hasEdge(self: *const Self, from: T, to: T) bool {
            const adjecent = self.adjacency.get(from);
            return adjecent != null and adjecent.?.contains(to);
        }

        pub fn addEdge(self: *Self, from: T, to: T, bidirectional: bool) !void {
            if (bidirectional) try self.addEdge(to, from, false);
            try self._vertices.put(from, {});
            try self._vertices.put(to, {});

            const entry = try self.adjacency.getOrPut(from);
            if (!entry.found_existing)
                entry.value_ptr.* = std.AutoArrayHashMap(T, i32).init(self._allocator);
            try entry.value_ptr.*.put(to, 0);
        }

        /// Slice of all outgoing nodes from this node
        pub fn outgoing(self: *const Self, node: T, allocator: std.mem.Allocator) ![]T {
            var nodes = std.ArrayList(T).init(allocator);

            if (self.adjacency.contains(node)) {
                for (self.adjacency.get(node).?.keys()) |trg| {
                    try nodes.append(trg);
                }
            }

            return nodes.toOwnedSlice();
        }

        /// True for nodes that have no outgoing edges
        pub fn isSink(self: *const Self, node: T) bool {
            return (!self.adjacency.contains(node) or self.adjacency.get(node).?.count() == 0);
        }

        /// Slice of all edges of graph
        pub fn edges(self: *const Self, allocator: std.mem.Allocator) ![][2]T {
            var edgelist = std.ArrayList([2]T).init(allocator);

            var node_iterator = self.adjacency.keyIterator();
            while (node_iterator.next()) |from| {
                for (self.adjacency.get(from.*).?.keys()) |to| {
                    try edgelist.append(.{ from.*, to });
                }
            }

            return edgelist.toOwnedSlice();
        }

        /// Inverting all edges
        pub fn invert(self: *Self, allocator: std.mem.Allocator) !Graph(T) {
            var inverted = Self.init(allocator);

            var node_iterator = self.adjacency.keyIterator();
            while (node_iterator.next()) |from| {
                for (self.adjacency.get(from.*).?.keys()) |to| {
                    try inverted.addEdge(to, from.*, false);
                }
            }

            return inverted;
        }

        /// Topological sorting with Kahn's algorithm
        pub fn topoSorting(self: *Self, allocator: std.mem.Allocator) ![]T {
            // Invert graph, because we need to check for node with no incoming edges
            var inverted = try self.invert(self._allocator);
            defer inverted.deinit();

            var sorted = std.ArrayList(T).init(allocator);

            // Start with all sinks (in inverted graph)
            var nodes = std.AutoArrayHashMap(T, void).init(allocator);
            defer nodes.deinit();
            for (inverted._vertices.keys()) |node| {
                if (inverted.isSink(node)) {
                    try nodes.put(node, {});
                }
            }

            // Loop will sinks left
            while (nodes.count() > 0) {
                // Add sink to list, remove their edges, find new sinks
                const n = nodes.pop().key;
                try sorted.append(n);

                const targets = try self.outgoing(n, allocator);
                defer allocator.free(targets);
                for (targets) |m| {
                    _ = inverted.adjacency.getPtr(m).?.swapRemove(n);
                    if (inverted.isSink(m)) {
                        try nodes.put(m, {});
                    }
                }
            }

            return sorted.toOwnedSlice();
        }
    };
}

pub fn PathFinding(comptime T: type) type {
    comptime std.debug.assert(@hasDecl(T, "Node"));
    comptime std.debug.assert(@hasDecl(T, "next"));
    comptime std.debug.assert(@hasDecl(T, "heuristic"));
    comptime std.debug.assert(@hasDecl(T, "isEnd"));

    return struct {
        base: *const T,

        const Self = @This();
        const OpenQueueItem = struct {
            f_score: u32,
            node: T.Node,

            fn lessThan(context: *std.AutoHashMap(T.Node, u32), a: OpenQueueItem, b: OpenQueueItem) math.Order {
                // Include g score in order, so that nodes that are closer to end are preferred
                const f_score_order = math.order(a.f_score, b.f_score);
                return f_score_order.differ() orelse math.order(context.get(a.node).?, context.get(b.node).?).invert();
            }

            pub fn format(self: @This(), comptime fmt: []const u8, options: std.fmt.FormatOptions, writer: anytype) !void {
                _ = fmt;
                _ = options;
                return std.fmt.format(writer, "({any}, {d})", .{ self.node, self.f_score });
            }
        };

        pub const Child = struct { node: T.Node, weight: u32 };

        pub const AStarResult = struct {
            cost: u32,
            reached_end_nodes: []T.Node,
            history: std.AutoHashMap(T.Node, std.ArrayList(T.Node)),

            /// Computes all paths that lead with minimal cost to an end state
            pub fn paths(self: *const @This(), allocator: std.mem.Allocator) ![][]T.Node {
                var path_list = std.ArrayList([]T.Node).init(allocator);

                // Create path for each end state
                for (self.reached_end_nodes) |end_node| {
                    const constructed_paths = try self.constructPaths(end_node, allocator);
                    try path_list.appendSlice(constructed_paths);
                    allocator.free(constructed_paths);
                }

                return path_list.toOwnedSlice();
            }

            /// Recursively constructs the paths with the history
            fn constructPaths(self: *const @This(), node: T.Node, allocator: std.mem.Allocator) ![][]T.Node {
                var path_list = std.ArrayList([]T.Node).init(allocator);
                if (self.history.get(node)) |prevs| {
                    // Take all previous nodes and create paths for each of them
                    for (prevs.items) |prev| {
                        const constructed_paths = try self.constructPaths(prev, allocator);
                        defer allocator.free(constructed_paths);
                        for (constructed_paths) |prev_path| {
                            defer allocator.free(prev_path);
                            var new_path = std.ArrayList(T.Node).init(allocator);
                            try new_path.appendSlice(prev_path);
                            try new_path.append(node);
                            try path_list.append(try new_path.toOwnedSlice());
                        }
                    }
                } else {
                    // No history, so this is the start node
                    var new_path = std.ArrayList(T.Node).init(allocator);
                    try new_path.append(node);
                    try path_list.append(try new_path.toOwnedSlice());
                }

                return try path_list.toOwnedSlice();
            }

            pub fn deinit(self: *@This(), allocator: std.mem.Allocator) void {
                allocator.free(self.reached_end_nodes);

                var values = self.history.valueIterator();
                while (values.next()) |val| val.deinit();
                self.history.deinit();
            }
        };

        /// AStar algorithm. Note that if neglect_other_paths is true, the paths construction in the result won't be complete.
        pub fn astar(self: *const Self, start: T.Node, neglect_other_paths: bool, allocator: std.mem.Allocator) !?AStarResult {
            // Link to previous nodes and list of reached final nodes for path tracking
            var history = std.AutoHashMap(T.Node, std.ArrayList(T.Node)).init(allocator);
            var end_nodes = std.ArrayList(T.Node).init(allocator);

            // Cost map
            var g_score_map = std.AutoHashMap(T.Node, u32).init(allocator);
            try g_score_map.put(start, 0);
            defer g_score_map.deinit();

            // Frontier of nodes
            var open = std.PriorityQueue(OpenQueueItem, *@TypeOf(g_score_map), OpenQueueItem.lessThan).init(allocator, &g_score_map);
            try open.add(.{ .f_score = 0, .node = start });
            defer open.deinit();

            // Best cost once found
            var best_path_cost: ?u32 = null;

            while (open.removeOrNull()) |current| {
                // Stop if lowest possible node already exceeds best cost solution (assumes admissible heuristic)
                if (best_path_cost != null and current.f_score > best_path_cost.?)
                    break;

                const g = g_score_map.get(current.node).?;
                // std.debug.print("{any} ({d} {d}) - {any}\n", .{ current.node, g, current.f_score, open.items });

                // Found the goal
                if (self.base.isEnd(current.node)) {
                    best_path_cost = best_path_cost orelse g;
                    try end_nodes.append(current.node);
                    if (neglect_other_paths) break;
                }

                // Append all successor nodes to frontier
                const children: []Child = try self.base.next(current.node, allocator);
                defer allocator.free(children);
                for (children) |child| {
                    const child_g = g + child.weight;

                    // If this child node is already evaluated, skip if this g value is worse
                    if (g_score_map.get(child.node)) |prev_child_g| {
                        if (prev_child_g <= child_g) {
                            // If g values are equal, we have multiple paths with same cost, so store both in history
                            if (prev_child_g == child_g) {
                                try history.getPtr(child.node).?.append(current.node);
                            }
                            continue;
                        }
                    }
                    try g_score_map.put(child.node, child_g);

                    // Compute heuristical value for frontier queue
                    const child_h: u32 = self.base.heuristic(child.node);
                    const child_f = child_g + child_h;
                    try open.add(.{ .f_score = child_f, .node = child.node });

                    // Since we have a better g value, update the history
                    if (history.getPtr(child.node)) |old_prev| old_prev.deinit();
                    var prevs = std.ArrayList(T.Node).init(allocator);
                    try prevs.append(current.node);
                    try history.put(child.node, prevs);
                }
            }

            // If best cost is still empty, goal was never found, otherwise create result
            if (best_path_cost) |cost| {
                return .{
                    .cost = cost,
                    .reached_end_nodes = try end_nodes.toOwnedSlice(),
                    .history = history,
                };
            } else {
                end_nodes.deinit();
                var values = history.valueIterator();
                while (values.next()) |val| val.deinit();
                history.deinit();
                return null;
            }
        }
    };
}
